File: scripts/stmt-proc-py/src/processors/hdfc_credit_card_processor.py
import pandas as pd
import hashlib
import os
from .statement_processor import StatementProcessor


class HdfcCreditCardStatementProcessor(StatementProcessor):
    """Processor for HDFC Credit Card Statements."""

    def __init__(self, txn_store):
        """Initialize with a transaction store."""
        super().__init__(txn_store)
        
    def statement_type(self):
        """Return the statement type."""
        return "hdfc-cc"

    def parse_statement(self, file_path):
        # Read the XLS file
        df = pd.read_excel(file_path, header=None)

        # Extract txn-source from the first 6 digits of the filename
        file_name = os.path.basename(file_path)
        txn_source = file_name[:6]

        # Find the header row where the 2nd column value is "Transaction type"
        header_row_index = df[df.iloc[:, 1] == "Transaction type"].index[0]
        start_row_index = header_row_index + 1
        df = df.iloc[start_row_index:].reset_index(drop=True)

        # Rename columns for easier access
        df.rename(columns={
            17: "txn_date",  # 18th column (0-based index is 17)
            21: "narration",  # 22nd column (0-based index is 21)
            48: "txn_amount",  # 49th column (0-based index is 48)
            54: "debit_credit"  # 55th column (0-based index is 54)
        }, inplace=True)

        # Process each transaction record
        transactions = []
        for _, row in df.iterrows():
            if pd.isna(row['txn_date']) or pd.isnull(row['txn_date']):
                break

            # Concatenate raw data
            raw_data = f"{row['txn_date']}|{row['narration']}|{row['txn_amount']}|{row['debit_credit']}"
            row_id = hashlib.md5(raw_data.encode()).hexdigest()

            # Parse transaction date
            txn_date = pd.to_datetime(row['txn_date'][:10], format='%d/%m/%Y').strftime('%Y-%m-%d')

            # Determine credit indicator
            credit_indicator = "Yes" if row['debit_credit'] == "Cr" else ""

            # Create transaction record
            transaction = {
                "row-id": row_id,
                "txn-source": txn_source,
                "txn-date": txn_date,
                "narration": row['narration'],
                "txn-amount": row['txn_amount'],
                "credit-indicator": credit_indicator,
                "txn-type": "",
                "category": "",
                "sub-category": "",
                "raw-data": raw_data,
            }
            transactions.append(transaction)

        # Store transactions in the consolidated CSV file
        self.store_transactions(transactions)

File: scripts/stmt-proc-wasm/src/hdfc_credit_card_processor.rs
// src/hdfc_credit_card_processor.rs
use std::error::Error;
use std::fs::File;
use std::path::Path;
use std::io::BufReader;
use calamine::{open_workbook, Reader, Xlsx};

use crate::statement_processor::{StatementProcessor, StatementProcessorBase, Transaction};
use crate::txn_store::TxnStore;

pub struct HdfcCreditCardStatementProcessor {
    base: StatementProcessorBase,
}

impl HdfcCreditCardStatementProcessor {
    pub fn new(txn_store: TxnStore) -> Self {
        Self {
            base: StatementProcessorBase::new(txn_store, "hdfc-cc".to_string()),
        }
    }
}

fn calculate_md5(data: &str) -> String {
    let digest = md5::compute(data);
    format!("{:x}", digest)
}

pub fn read_excel(file_path: &str) -> Result<(), Box<dyn std::error::Error>> {
    let mut workbook: Xlsx<_> = open_workbook(file_path)?;
    if let Some(Ok(range)) = workbook.worksheet_range("Sheet1") {
        for row in range.rows() {
            println!("{:?}", row);
        }
    }
    Ok(())
}

impl StatementProcessor for HdfcCreditCardStatementProcessor {

    fn statement_type(&self) -> String {
        self.base.statement_type().to_string()
    }

    fn txn_store(&self) -> &TxnStore {
        &self.base.txn_store()
    }

    fn store_transactions(&self, transactions: Vec<Transaction>) {
        self.base.store_transactions(transactions);
    }

    fn parse_statement(&self, file_path: &str) -> Result<(), Box<dyn Error>> {
        let file = File::open(file_path)?;
        let reader = BufReader::new(file);
        let mut rdr = csv::ReaderBuilder::new().has_headers(false).from_reader(reader);

        let txn_source = Path::new(file_path)
            .file_name()
            .unwrap()
            .to_str()
            .unwrap()[..6]
            .to_string();

        let mut transactions: Vec<Transaction> = Vec::new();

        for result in rdr.records() {
            let record = result?;
            let txn_date = &record[17]; // 18th column
            let narration = &record[21]; // 22nd column
            let txn_amount: f64 = record[48].parse()?;
            let debit_credit = &record[54]; // 55th column

            if txn_date.is_empty() {
                break;
            }

            let raw_data = format!("{}|{}|{}|{}", txn_date, narration, txn_amount, debit_credit);
            let row_id = calculate_md5(&raw_data);

            let txn_date = txn_date[..10].to_string(); // Assuming date format is correct
            let credit_indicator = if debit_credit == "Cr" { "Yes" } else { "" };

            let transaction = Transaction {
                row_id,
                txn_source: txn_source.to_string(),
                txn_date,
                narration: narration.to_string(),
                txn_amount,
                credit_indicator: credit_indicator.to_string(),
                txn_type: "".to_string(),
                category: "".to_string(),
                sub_category: "".to_string(),
                raw_data,
            };

            transactions.push(transaction);
        }

        self.store_transactions(transactions);
        Ok(())
    }
}